src/dataflow/src/storage/render/debezium.rs

```diff
@@ -118,6 +121,304 @@ where
+                                        tx_cap_map
+                                            .insert(tx_id.clone(), tx_metadata_cap.clone());
+                                    }
+                                    Some(val) if val == time => {},
+                                    Some(val) => panic!("unexpected mismatch in duplicate END record for {:?}: {:?} vs {:?}", tx_id, time, val),
+                                }
+                            }
+                        }
+
+                        while let Some((data_cap, refmut_data)) = data_input.next() {
+                            // RefOrMut doesn't let us drain directly into iterator
+                            refmut_data.swap(&mut data);
+                            let data_cap = data_cap.retain();
+                            data_buffer.extend(data.drain(..).map(|r| (r, data_cap.clone())));
+                        }
+                        while let Some((result, data_cap)) = data_buffer.get(0).cloned() {
+                            let key = match result.key.transpose() {
+                                Ok(key) => key,
+                                Err(err) => {
+                                    output.session(&data_cap).give((
+                                        Err(err.into()),
+                                        *data_cap.time(),
+                                        1,
+                                    ));
+                                    let _ = data_buffer.pop_front();
+                                    continue;
+                                }
+                            };
+                            let value = match result.value {
+                                Some(Ok(value)) => value,
+                                Some(Err(err)) => {
+                                    output.session(&data_cap).give((
+                                        Err(err.into()),
+                                        *data_cap.time(),
+                                        1,
+                                    ));
+                                    let _ = data_buffer.pop_front();
+                                    continue;
+                                }
+                                None => {
+                                    let _ = data_buffer.pop_front();
+                                    continue;
+                                }
+                            };
+
+                            let tx_id = match value.iter().nth(transaction_idx).unwrap() {
+                                Datum::List(l) => {
+                                    let tx_id = l.iter().nth(tx_metadata_description.data_transaction_id_idx).unwrap().unwrap_str().to_owned();
+                                    let tx_time: Timestamp = match tx_mapping.get(&tx_id) {
+                                        Some(time) => *time,
+                                        None => break,
```

Comment:
This is the one control flow path where the element stays in the deque and is not popped so I would flip the logic around and pop the element as part of the while loop and only put it back here right before breaking.

Here is a proposed diff:

```diff
diff --git a/src/dataflow/src/storage/render/debezium.rs b/src/dataflow/src/storage/render/debezium.rs
index bc1cd13e6..3f8bad83b 100644
--- a/src/dataflow/src/storage/render/debezium.rs
+++ b/src/dataflow/src/storage/render/debezium.rs
@@ -269,8 +269,8 @@ where
                             let data_cap = data_cap.retain();
                             data_buffer.extend(data.drain(..).map(|r| (r, data_cap.clone())));
                         }
-                        while let Some((result, data_cap)) = data_buffer.get(0).cloned() {
-                            let key = match result.key.transpose() {
+                        while let Some((result, data_cap)) = data_buffer.pop_front() {
+                            let key = match result.key.clone().transpose() {
                                 Ok(key) => key,
                                 Err(err) => {
                                     output.session(&data_cap).give((
@@ -278,33 +278,35 @@ where
                                         *data_cap.time(),
                                         1,
                                     ));
-                                    let _ = data_buffer.pop_front();
                                     continue;
                                 }
                             };
-                            let value = match result.value {
-                                Some(Ok(value)) => value,
+                            let value = match result.value.clone() {
+                                Some(Ok(value)) => value.clone(),
                                 Some(Err(err)) => {
                                     output.session(&data_cap).give((
                                         Err(err.into()),
                                         *data_cap.time(),
                                         1,
                                     ));
-                                    let _ = data_buffer.pop_front();
                                     continue;
                                 }
                                 None => {
-                                    let _ = data_buffer.pop_front();
                                     continue;
                                 }
                             };

                             let tx_id = match value.iter().nth(transaction_idx).unwrap() {
                                 Datum::List(l) => {
-                                    let tx_id = l.iter().nth(tx_metadata_description.data_transaction_id_idx).unwrap().unwrap_str().to_owned();
+                                    let tx_id = l.iter().nth(data_tx_id_idx).unwrap().unwrap_str().to_owned();
                                     let tx_time: Timestamp = match tx_mapping.get(&tx_id) {
                                         Some(time) => *time,
-                                        None => break,
+                                        None => {
+                                            // We don't yet have a time for this piece of data so
+                                            // bail until we get more transaction metadata
+                                            data_buffer.push_front((result, data_cap));
+                                            break;
+                                        }
                                     };
                                     Some((tx_id, tx_time))
                                 },
@@ -312,9 +314,6 @@ where
                                 _ => panic!("Previously validated to be nullable list"),
                             };

-                            // We're committed to processing it if we have a tx_time
-                            let _ = data_buffer.pop_front();
-
                             let partition_dedup = dedup_state
                                 .entry(result.partition.clone())
                                 .or_insert_with(|| {
```